engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala

```diff
@@ -1187,59 +1187,74 @@ class IrToTruffle(
       * @return a node for the final shape of function body and pre-processed
       *         argument definitions.
       */
-    def buildFunctionBody(
-      arguments: List[IR.DefinitionArgument],
-      body: IR.Expression
-    ): (BlockNode, Array[ArgumentDefinition]) = {
+    class BuildFunctionBody(
+      val arguments: List[IR.DefinitionArgument],
+      val body: IR.Expression
+    ) {
       val argFactory = new DefinitionArgumentProcessor(scopeName, scope)
+      private var argDefinitions: Array[ArgumentDefinition] = null
```

Comment:
I think the scala way would be:
```
private lazy val (argDefinitions, argSlots, argExpressions) = slots()
```
Then you won't need the `if (argSlots == null)` inside `slots()`, just write the code that computes it. And anything that accesses the lazy fields will automatically force `slots()` to run, exactly once.

Reply:
I knew there is going to be something fancy in Scala to achieve that! Done in f29ce903cf1d283 and yes, it is nicer than my original Javaish attempt.